File: sdks/swift/rest/Sources/Schemas/PublishResponse.swift
import Foundation

/// `POST /v1/channels/:channel/publish` success response.
public struct PublishResponse: Codable, Hashable, Sendable {
    public let channel: String
    public let event: String
    /// Number of subscribers the message was fanned out to.
    public let recipients: Int
    /// Always `"published"`. Reserved for future state changes.
    public let status: String
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        channel: String,
        event: String,
        recipients: Int,
        status: String,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.channel = channel
        self.event = event
        self.recipients = recipients
        self.status = status
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.channel = try container.decode(String.self, forKey: .channel)
        self.event = try container.decode(String.self, forKey: .event)
        self.recipients = try container.decode(Int.self, forKey: .recipients)
        self.status = try container.decode(String.self, forKey: .status)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encode(self.channel, forKey: .channel)
        try container.encode(self.event, forKey: .event)
        try container.encode(self.recipients, forKey: .recipients)
        try container.encode(self.status, forKey: .status)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case channel
        case event
        case recipients
        case status
    }
}

File: sdks/rust/rest/src/api/types/publish_response.rs
pub use crate::prelude::*;

/// `POST /v1/channels/:channel/publish` success response.
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq, Hash)]
pub struct PublishResponse {
    #[serde(default)]
    pub channel: String,
    #[serde(default)]
    pub event: String,
    /// Number of subscribers the message was fanned out to.
    #[serde(default)]
    pub recipients: i64,
    /// Always `"published"`. Reserved for future state changes.
    #[serde(default)]
    pub status: String,
}

impl PublishResponse {
    pub fn builder() -> PublishResponseBuilder {
        <PublishResponseBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct PublishResponseBuilder {
    channel: Option<String>,
    event: Option<String>,
    recipients: Option<i64>,
    status: Option<String>,
}

impl PublishResponseBuilder {
    pub fn channel(mut self, value: impl Into<String>) -> Self {
        self.channel = Some(value.into());
        self
    }

    pub fn event(mut self, value: impl Into<String>) -> Self {
        self.event = Some(value.into());
        self
    }

    pub fn recipients(mut self, value: i64) -> Self {
        self.recipients = Some(value);
        self
    }

    pub fn status(mut self, value: impl Into<String>) -> Self {
        self.status = Some(value.into());
        self
    }

    /// Consumes the builder and constructs a [`PublishResponse`].
    /// This method will fail if any of the following fields are not set:
    /// - [`channel`](PublishResponseBuilder::channel)
    /// - [`event`](PublishResponseBuilder::event)
    /// - [`recipients`](PublishResponseBuilder::recipients)
    /// - [`status`](PublishResponseBuilder::status)
    pub fn build(self) -> Result<PublishResponse, BuildError> {
        Ok(PublishResponse {
            channel: self
                .channel
                .ok_or_else(|| BuildError::missing_field("channel"))?,
            event: self
                .event
                .ok_or_else(|| BuildError::missing_field("event"))?,
            recipients: self
                .recipients
                .ok_or_else(|| BuildError::missing_field("recipients"))?,
            status: self
                .status
                .ok_or_else(|| BuildError::missing_field("status"))?,
        })
    }
}
